scylla.go

```diff
@@ -4,17 +4,96 @@ import (
 	"fmt"
 	"math"
 	"strconv"
+	"strings"
 	"sync/atomic"
 )
 
 // scyllaSupported represents Scylla connection options as sent in SUPPORTED
 // frame.
+// FIXME: Should also follow `cqlProtocolExtension` interface.
 type scyllaSupported struct {
 	shard             int
 	nrShards          int
 	msbIgnore         uint64
 	partitioner       string
 	shardingAlgorithm string
+	lwtFlagMask       int
+}
+
+// CQL Protocol extension interface for Scylla.
+// Each extension is identified by a name and defines a way to serialize itself
+// in STARTUP message payload.
+type cqlProtocolExtension interface {
+	name() string
+	serialize() map[string]string
+}
+
+func findCQLProtoExtByName(exts []cqlProtocolExtension, name string) cqlProtocolExtension {
+	for i := range exts {
+		if exts[i].name() == name {
+			return exts[i]
+		}
+	}
+	return nil
+}
+
+// Top-level keys used for serialization/deserialization of CQL protocol
+// extensions in SUPPORTED/STARTUP messages.
+// Each key identifies a single extension.
+const (
+	lwtAddMetadataMarkKey = "SCYLLA_LWT_ADD_METADATA_MARK"
+)
+
+// "LWT prepared statements metadata mark" CQL Protocol Extension.
+// This extension, if enabled (properly negotiated), allows Scylla server
+// to set a special bit in prepared statements metadata, which would indicate
+// whether the statement at hand is LWT statement or not.
+//
+// This is further used to consistently choose primary replicas in a predefined
+// order for these queries, which can reduce contention over hot keys and thus
+// increase LWT performance.
+//
+// Implements cqlProtocolExtension interface.
+type lwtAddMetadataMarkExt struct {
+	lwtOptMetaBitMask int
+}
+
+// Factory function to deserialize and create an `lwtAddMetadataMarkExt` instance
+// from SUPPORTED message payload.
+func newLwtAddMetaMarkExt(supported map[string][]string) *lwtAddMetadataMarkExt {
```

Comment:
Why return a pointer here when you de reference it right away at the call site?

Reply:
What am I supposed to return from the function in case I cannot construct a proper `lwtAddMetadataMarkExt` object?

In this case I can return `nil` and employ a straightforward logic of checking for `nil` to assert that the object is ok.

In your case, AFAIU, it means that I would construct an `lwtAddMetadataMarkExt` with empty name, which I cannot do, if I am not to break the invariant of this type. 
Another approach is to either return an empty `lwtAddMetadataMarkExt{}` and check if the `lwtOptMetaBitMask` is empty (seems a little bit hackish) or return a pair (ext, error) and check for error instead (but it's not really an error if we could not find the extension key in `supported` map, so I don't think this would be good).

I am personally in favor of either the very first option or the last one

Reply:
But if it's something with "don't return pointers to structs/interfaces unless you really-really have to" rule, I can adjust.

Reply:
This type is very small

```
type lwtAddMetadataMarkExt struct {
	lwtOptMetaBitMask int
}
```

Plus in L90 it's `lwtAddMetadataMarkExt` that implements the interface not `*lwtAddMetadataMarkExt` so from this pov you could return `lwtAddMetadataMarkExt`.

Note that just the fact of having it accessed over interface would force compiler to escape that to heap.

My suggestion would be to change it so that `*lwtAddMetadataMarkExt` satisfies interface and leave this line as is.

Reply:
Hmm, it seems that I cannot duplicate implementing methods with both value and pointer receivers. So I assume it's easier to convert this function to return value intstead of a pointer, because otherwise it would require to change a lot more code.

Reply:
You cannot but compiler should do the work for you so there should be no code changes i.e. you can run function with a ptr receiver on struct.

Reply:
I think in that case I would need to adjust type assertion at the struct declaration to the following form:
```go
var _ cqlProtocolExtension = &lwtAddMetadataMarkExt{}
```
Which would effectively mean that I have a dangling object on the heap which cannot be recycled?
Or is the compiler smart enough to produce no code for this case?

Reply:
I believe so since the `_` name makes unavailable for referencing.

Reply:
@dahankzter @mmatczuk Done.